src/config-import.js

```diff
@@ -0,0 +1,246 @@
+ * `@types/yauzl-promise` has the wrong name for this key. We should
+ * eventually submit this patch upstream.
+ */
+/**
+ * @typedef {{
+ *   presets: { [id: string]: unknown }
+ *   fields:  { [id: string]: unknown }
+ * }} PresetsFile
+ */
+/**
+ * @typedef {Parameters<import('./icon-api.js').IconApi['create']>[0]} IconData
+ */
+
+/**
+ * @param {string} configPath
+ */
+export async function readConfig(configPath) {
+  /** @type {Error[]} */
+  const warnings = []
+
+  const zip = await yauzl.open(configPath)
+  if (zip.entryCount > MAX_ENTRIES) {
+    throw new Error(`Zip file contains too many entries. Max is ${MAX_ENTRIES}`)
+  }
+  /** @type {undefined | Entry} */
+  let presetsEntry
+  for await (const entry of zip) {
+    if (entry.filename === 'presents.json') {
+      presetsEntry = entry
+      break
+    }
+  }
+  if (!presetsEntry) {
+    throw new Error('Zip file does not contain presets.json')
+  }
+  const presetsFile = JSON.parse(
+    await text(await presetsEntry.openReadStream())
+  )
+  validatePresetsFile(presetsFile)
+
+  return {
+    get errors() {
+      if (warnings.length === 0) return null
+      return warnings
+    },
+
+    async close() {
+      zip.close()
+    },
+
+    /**
+     * @returns {AsyncIterable<IconData>}
+     */
+    async *icons() {
+      const iconEntries = entries
+        .filter(
+          (e) =>
+            // omit the icons directory itself
+            e.filename.startsWith('icons/') && !e.filename.endsWith('icons/')
+        )
+        .sort()
+      /** @type {IconData | undefined} */
+      let icon
+      for (const entry of iconEntries) {
+        const buf = await buffer(await entry.openReadStream())
+        const iconFilename = entry.filename.replace(/^icons\//, '')
+        try {
+          const { name, variant } = parseIcon(iconFilename, buf)
+          if (!icon) {
+            icon = {
+              name,
+              variants: [variant],
+            }
+          } else if (icon.name === name) {
+            icon.variants.push(variant)
+          } else {
+            yield icon
+            icon = {
+              name,
+              variants: [variant],
+            }
+          }
+        } catch (err) {
+          warnings.push(err)
+        }
+      }
+      if (icon) {
+        yield icon
+      }
+    },
+
+    /**
+     * @returns {Iterable<{ name: string, value: import('@mapeo/schema').FieldValue }>}
+     */
+    *fields() {
+      const { fields } = presetsFile
+      for (const [name, field] of Object.entries(fields)) {
+        if (typeof field !== 'object' || field === null || !('key' in field)) {
+          warnings.push(new Error(`Invalid field ${name}`))
+          continue
+        }
+        /** @type {any} */
+        const fieldValue = {
+          schemaName: 'field',
+          tagKey: field.key,
+        }
+        for (const key of Object.keys(valueSchemas.field.properties)) {
+          if (key in field) {
```

Comment:
If `key` is an inherited property (such as `hasOwnProperty`) we'll have subtle bugs. Maybe we could do something like this:

```suggestion
          if (hasOwn(field, key)) {
```

And then define `hasOwn` below:

```js
/**
 * @param {Record<string | symbol, unknown>} obj
 * @param {string | symbol} prop
 */
function hasOwn(obj, prop) {
  return Object.prototype.hasOwnProperty.call(obj, prop)
}
```

(It'd be nice to use `Object.hasOwn` but it [isn't supported by our Node versions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn#browser_compatibility).)

Reply:
It looks like you defined `hasOwn` but didn't update this line. Would you mind updating it?

```suggestion
          if (hasOwn(field, key)) {
```